src/ocean_data_assim/MOM_oda_incupd.F90

```diff
@@ -0,0 +1,874 @@
+!! 5. output_oda_incupd_inc (output increment if using full fields input)
+!! 6. init_oda_incupd_diags (to output increments in diagnostics)
+!! 7. oda_incupd_end (not being used for now)
+
+module MOM_oda_incupd
+
+
+! This file is part of MOM6. See LICENSE.md for the license.
+use MOM_array_transform, only : rotate_array
+use MOM_coms,            only : sum_across_PEs
+use MOM_diag_mediator,   only : post_data, query_averaging_enabled, register_diag_field
+use MOM_diag_mediator,   only : diag_ctrl
+use MOM_domains,         only : pass_var,pass_vector
+use MOM_error_handler,   only : MOM_error, FATAL, NOTE, WARNING, is_root_pe
+use MOM_file_parser,     only : get_param, log_param, log_version, param_file_type
+use MOM_get_input,       only : directories, Get_MOM_input
+use MOM_grid,            only : ocean_grid_type
+use MOM_io,              only : vardesc, var_desc
+use MOM_remapping,       only : remapping_cs, remapping_core_h, initialize_remapping
+use MOM_restart,         only : register_restart_field, register_restart_pair, MOM_restart_CS
+use MOM_restart,         only : restart_init, save_restart, query_initialized
+use MOM_spatial_means,   only : global_i_mean
+use MOM_time_manager,    only : time_type
+use MOM_unit_scaling,    only : unit_scale_type
+use MOM_variables,       only : thermo_var_ptrs
+use MOM_verticalGrid,    only : verticalGrid_type
+use MOM_verticalGrid,    only : get_thickness_units
+
+use mpp_io_mod, only : mpp_get_axis_length
+use mpp_io_mod, only : axistype
+
+implicit none ; private
+
+#include <MOM_memory.h>
+
+
+!  Publicly available functions
+public set_up_oda_incupd_field, set_up_oda_incupd_vel_field
+public initialize_oda_incupd_fixed, initialize_oda_incupd, apply_oda_incupd, oda_incupd_end
+public init_oda_incupd_diags,calc_oda_increments,output_oda_incupd_inc
+
+! A note on unit descriptions in comments: MOM6 uses units that can be rescaled for dimensional
+! consistency testing. These are noted in comments with units like Z, H, L, and T, along with
+! their mks counterparts with notation like "a velocity [Z T-1 ~> m s-1]".  If the units
+! vary with the Boussinesq approximation, the Boussinesq variant is given first.
+
+!> A structure for creating arrays of pointers to 3D arrays with extra gridding information
+type :: p3d
+  integer :: id !< id for FMS external time interpolator
+  integer :: nz_data !< The number of vertical levels in the input field.
+  real, dimension(:,:,:), pointer :: mask_in => NULL() !< pointer to the data mask.
+  real, dimension(:,:,:), pointer :: p => NULL() !< pointer to the data.
+  real, dimension(:,:,:), pointer :: h => NULL() !< pointer to the data grid.
+end type p3d
+
+!> oda incupd control structure
+type, public :: oda_incupd_CS ; private
+  integer :: nz        !< The total number of layers.
+  integer :: nz_data   !< The total number of arbritary layers (used by older code).
+  integer :: isc       !< The starting i-index of the computational domain at h.
```

Comment:
The 16 horizontal index bounds (CS%isc, CS%iec, ...) in this control structure are set but never used. Instead the values from the ocean grid type are used (appropriately). Why are these elements included in this control structure, and could they be removed without loss?

Reply:
yes, I just tested that they can be removed from the control structure as well as their initialization in MOM_oda_incupd.F90/initialize_oda_incupd. @jiandewang could you have Phil correct that too?

Reply:
Thanks Alex, I will let Phil know this